dcicutils/misc_utils.py

```diff
@@ -1630,16 +1631,140 @@ class SubFreeze(Freeze):
+        """
+        Clears the cache for the given attribute name on the given instance_class.
+
+        Having done:
+
+            import random
+            class Foo:
+                @classproperty_cached
+                def something(cls):
+                    return random.randint(100)
+
+        one can clear the cache by doing:
+
+            classproperty_cached.reset(instance_class=Foo, attribute_name='something')
+
+        :param instance_class: a class with an attribute whose value is managed by '@classproperty_cached'.
+        :param attribute_name: the name of the attribute that has a cached value.
+        :param subclasses: a bool indicating whether the cache reset requests applies to
+             all subclasses (subclasses=True) or only the exact class given (subclasses=False).
+             Using subclasses=False is not allowed if CACHE_EACH_SUBCLASS is False.
+        """
+
+        if not cls._USES_PER_SUBCLASS_CACHES and not subclasses:
+            raise ValueError(f"The subclasses= argument to {cls.__name__}.reset must not be False"
+                             f" because {cls.__name__} does not use per-subclass caches.")
+        if not isinstance(instance_class, type):
+            raise ValueError(f"The instance_class= argument to {cls.__name__}.reset must be a class.")
+        reference_class = cls._find_reference_class(instance_class, attribute_name)
+        attribute_cache = cls._find_cache_map(reference_class, attribute_name)
+        keys_to_remove = []
+        predicate = issubclass if subclasses else equals
+        for cache_class, cache_value in attribute_cache.items():
+            if predicate(cache_class, reference_class):
+                keys_to_remove.append(cache_class)
+        for cache_class in keys_to_remove:
+            del attribute_cache[cache_class]
+        return bool(keys_to_remove)
+
+
+class classproperty_cached_each_subclass(classproperty_cached):
+    """
+    This decorator is like 'classproperty', but the function is run only on first use, not every time, and then cached.
+
+    Such a property returns the same value each time, just like any class property,
+    but initialization is delayed until first use and might be the result of
+    a multi-line computation whose temporary variables don't persist past the initialization.
+
+    Example:
+
+        import time
+        class Freeze:
+            @classproperty_cached
+            def sample():
+                return time.time()
+
+        Freeze.sample
+        1665600374.4801269
+        Freeze.sample
+        1665600374.4801269
+
+        class SubFreeze(Freeze):
+            pass
+
+        SubFreeze.sample
+        1665600540.1467211
+        SubFreeze.sample
+        1665600540.1467211
+
+        Freeze.sample
+        1665600374.4801269
+
+        SubFreeze.sample
+        1665600540.1467211
+    """
+
+    _USES_PER_SUBCLASS_CACHES = True
+
+
+def equals(x, y):
```

Comment:
If the only usage of this is in the class why not just do `lambda x, y: x == y` and avoid the function call/definition?

Reply:
I find these things useful from time to time and it annoys me coming from a language like Lisp that doesn't have an "operator" distinction not to have this available as a function, so it seemed that it would enable other uses prospectively in the future.